code/test/tokens.js

```diff
@@ -0,0 +1,501 @@
+/* TODO
+1. Remove TestUtil and merge with util.js
+2. Return oldState when calling makeWallet and assetTransfer
+3. Retrieve currentState as part of checkONEWalletStateChange
+3. Remove commented lines
+4. Use batch function for token Tracker
+5. Add TRACK and UNTRACK and OVERRIDE_TRACK Tests
+6. Enhance assetTransfer to be walletTransaction (and cater for additional operations)
+*/
+
+const TestUtil = require('./util')
+const unit = require('ethjs-unit')
+const ONEUtil = require('../lib/util')
+const ONEConstants = require('../lib/constants')
+// const BN = require('bn.js')
+
+const ONE_CENT = unit.toWei('0.01', 'ether')
+// const HALF_DIME = unit.toWei('0.05', 'ether')
+// const ONE_DIME = unit.toWei('0.1', 'ether')
+const HALF_ETH = unit.toWei('0.5', 'ether')
+const INTERVAL = 30000 // 30 second Intervals
+const DURATION = INTERVAL * 12 // 6 minute wallet duration
+// const SLOT_SIZE = 1 // 1 transaction per interval
+const EFFECTIVE_TIME = Math.floor(Date.now() / INTERVAL / 6) * INTERVAL * 6 - DURATION / 2
+const DUMMY_HEX = ONEUtil.hexString('5') // Dummy Hex string for 5 i.e. 0x05
+
+contract('ONEWallet', (accounts) => {
+  // Wallets effective time is the current time minus half the duration (3 minutes ago)
+  let snapshotId
+  beforeEach(async function () {
+    snapshotId = await TestUtil.snapshot()
+    await TestUtil.init()
+  })
+  afterEach(async function () {
+    await TestUtil.revert(snapshotId)
+  })
+
+  // Test creation and validation of wallet
+  it('Wallet_Validate: must be able to create and validate a wallet', async () => {
+    let alice = await TestUtil.makeWallet('TT-WALLET-1', accounts[0], EFFECTIVE_TIME, DURATION)
+    let aliceOldState = await TestUtil.getONEWalletState(alice.wallet)
+    let aliceCurrentState = await TestUtil.getONEWalletState(alice.wallet)
+    await TestUtil.checkONEWalletStateChange(aliceOldState, aliceCurrentState)
+  })
+
+  // Transfer Native Asset to external wallet
+  it('Wallet_CommitReveal: Native Asset Transfer must commit and reveal successfully', async () => {
```

Comment:
For each test, let's only focus on validating specific feature, rather than everything related to how the wallet should function as of this version - the wallet's interfaces, features, and expected functions may change across versions and over time. When they change, we don't want to update all tests in all files. We only want to modify one test specifically relevant to that feature, and change a small part of the code in a specific file.

Simple transfer is already tested in basic - let's remove it from here and only tests for tokens. Similarly Wallet_Validate can be removed